### PR TITLE
Fix User Defined Index Selection

### PR DESCRIPTION
**Pull request by tonysun83**
Users were complaining that certain operators such as $in and $size were throwing a "no usable index" error. It turns out that when a user defines their own set of fields to be indexed, we weren't correctly selecting the indexes because of our field generator. Example, when a user did

{"selector": "favorites": {"$in": ["Ruby"]}}

We actually create two fields underneath the covers: favorites.[]:string, and favorites:string. We need these extra fields to correctly perform certain searches such as $in.

When users use the operator $size, we also create favorites.[]:length.

When users index an array, they usually just index one field, i.e favorites.[]:string. When selecting an index, we make sure the fields from the selector are a subset of the user-defined fields. Unfortunately, because we generate extra fields, this throws the "no usable_index error". 

This fix modifies indexable_fields to remove extraneous fields (only for the fields check, not from the actual selector) that are created by the mango_selector_text:convert function so that
the index can now be used when accessing arrays, $in operations, and
$size operations.

### CODE REVIEW

File: src/mango_idx_text.erl [hunk 1]
```diff
@@ -283,6 +283,23 @@ indexable_fields(Fields, {op_and, Args}) when is_list(Args) ->
...
+indexable_fields(Fields, {op_or, [{op_field, Field0},
+        {op_field, {[Name | _], _}} = Field1]}) ->
+    case lists:member(<<"[]">>, Name) of
+        true ->
+            indexable_fields(Fields, Field1);
+        false ->
+            Fields1 = indexable_fields(Fields, Field0),
+            indexable_fields(Fields1, Field1)
+    end;
+indexable_fields(Fields, {op_or, Args}) when is_list(Args) ->
```
Comment (kxepal) [inline src/mango_idx_text.erl]: What's the difference between this function clause and the one below?

Comment (tonysun83) [inline src/mango_idx_text.erl]: @kxepal : The first function clause addresses two unique situations.

First is when the user does an $in operation with a list. Each element in the list will construct two fields with an op_or:
https://github.com/apache/couchdb-mango/blob/master/src/mango_selector_text.erl#L241-L260

The second situation is when the user access individual elements in the list. Again, there are two fields created with an op_or:
https://github.com/apache/couchdb-mango/blob/master/src/mango_selector_text.erl#L184-L191

In both these situations, the second op_field will contain a "[]" reference, which is a special identifier for that says we are referencing an array. So you may ask? What if a user asks

Give me all documents where the favorites field is equal to 5 or the first element a favorites array is 10.

"$or": [{"favorites": 5}, {"favorites.0": 10}]

This actually creates a nested op_or and will bypass the first function clause. Then the first "favorites": 5 will just be an op_field, while the second   {"favorites.0": 10} will be create an op_or that will be handled by the first function clause. 

So this way, we fix $in, and array element access without breaking $or.

Comment (kxepal) [inline src/mango_idx_text.erl]: @tonysun83 I actually mean that lines 300-302 duplicates lines 303-305 (:

Comment (tonysun83) [inline src/mango_idx_text.erl]: Oh derp derp, lol, thx for noticing that.

Comment (kxepal) [inline src/mango_idx_text.erl]: (:

**Comment by kxepal** [comment]
+1 after squash.

**Comment by kxepal** [comment]
Btw, need to find a way to run text index related tests on travis. May be we can maintain some patchset to install dreyfus and related things? cc @robertkowalski

**Comment by tonysun83** [comment]
Agreed, I'll get to this after (hopefully) improving our mango performance.

**Pull request merged**